### PR TITLE
shim notifier

### PR DESCRIPTION
**Pull request by wanqizhu**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
https://improbableio.atlassian.net/browse/UNR-482

Making a function 'AActor::OnSpatialAuthorityChange' which gets triggered inside `SpatialPipelineBlock::LeaveCriticalSection`. This can be used to process handover properties, especially once we implement soft boundaries.

Will be synced with an Engine change. https://github.com/improbableio/UnrealEngine/pull/3

#### Tests
Running across worker boundaries work as normal. Logging reveals that by the time OnSpatialAuthorityChange is called, all handover and replicated properties have already been processed. 

Right now, this function is also only invoked once per actor on handover and only on the receiving worker (the worker with authority, so AuthorityOp is always 1). This is because we only trigger this as we leave critical section and it seems that by then, the old server worker already lost authority.

This is fine for supporting handover notification. A more robust change will be to pass in the AuthorityOp into `OnSpatialAuthorityChange` and trigger it from multiple places -- not sure if that's necessary as we don't know what the use case would be. This shim is just to spike the concept.

#### Documentation
n/a -- later on, we may want to provide an example use case for `OnSpatialAuthorityChange()`

#### Primary reviewers
@m-samiec @Vatyx

### CODE REVIEW

File: Source/SpatialGDK/Private/SpatialInteropPipelineBlock.cpp [hunk 1]
```diff
@@ -187,6 +187,21 @@ void USpatialInteropPipelineBlock::LeaveCriticalSection()
 		RemoveEntityImpl(PendingRemoveEntity);
 	}
 
+	// Notify authority changes
+	for (auto& PendingAuthorityChange : PendingAuthorityChanges)
+	{
+		// Use position component as a proxy for overall actor authority
+		if (PendingAuthorityChange.Key.ComponentId == UPositionComponent::ComponentId)
```
Comment (Vatyx) [inline Source/SpatialGDK/Private/SpatialInteropPipelineBlock.cpp]: You can use the `View` to directly get authority rather than using another component.

Comment (wanqizhu) [reply Source/SpatialGDK/Private/SpatialInteropPipelineBlock.cpp]: Chatted offline -- current implementation is okay.

File: Source/SpatialGDK/Private/SpatialInteropPipelineBlock.cpp [hunk 2]
```diff
@@ -187,6 +187,21 @@ void USpatialInteropPipelineBlock::LeaveCriticalSection()
...
+		{
+			AActor* EntityActor = EntityRegistry->GetActorFromEntityId(PendingAuthorityChange.Key.EntityId);
```
Comment (m-samiec) [inline Source/SpatialGDK/Private/SpatialInteropPipelineBlock.cpp]: Move actor definition into if statement.

File: Source/SpatialGDK/Private/SpatialInteropPipelineBlock.cpp [hunk 3]
```diff
@@ -187,6 +187,21 @@ void USpatialInteropPipelineBlock::LeaveCriticalSection()
 		RemoveEntityImpl(PendingRemoveEntity);
 	}
 
+	// Notify authority changes
+	for (auto& PendingAuthorityChange : PendingAuthorityChanges)
```
Comment (m-samiec) [inline Source/SpatialGDK/Private/SpatialInteropPipelineBlock.cpp]: Move this into AuthorityChangeImpl and call it from ChangeAuthority() when we're not in a critical section.

Comment (wanqizhu) [reply Source/SpatialGDK/Private/SpatialInteropPipelineBlock.cpp]: Good point, will do.

File: Source/SpatialGDK/Private/SpatialInteropPipelineBlock.cpp (hunk 2 above)
Comment (wanqizhu) [reply Source/SpatialGDK/Private/SpatialInteropPipelineBlock.cpp]: done

**Pull request merged**